File: 2024/day1.go
package main

import (
	"math"
	"slices"
	"strconv"
	"strings"
)

func (d *Day) Day1(input []string) {
	leftList := make([]int, len(input))
	rightList := make([]int, len(input))

	// Parsing
	for i, line := range input {
		columns := strings.Fields(line)
		if len(columns) != 2 {
			panic("There should be two columns!")
		}
		left, err := strconv.Atoi(columns[0])
		if err != nil {
			panic(err)
		}
		right, err := strconv.Atoi(columns[1])
		if err != nil {
			panic(err)
		}
		leftList[i] = left
		rightList[i] = right
	}

	// Sorting
	slices.Sort(leftList)
	slices.Sort(rightList)

	// Part 1
	var sum int
	for i, leftNum := range leftList {
		sum += int(math.Abs(float64(leftNum - rightList[i])))
	}

	d.SetPart1Answer(sum)

	// Part 2
	sum = 0
	for _, leftNum := range slices.Compact(leftList) {
		var occurrences int
		for _, rightNum := range rightList {
			if leftNum > rightNum {
				continue
			}
			if leftNum == rightNum {
				occurrences++
				continue
			}
			break
		}
		sum += leftNum * occurrences
	}

	d.SetPart2Answer(sum)
}

File: 2024/src/main/kotlin/days/Day1.kt
package days

import kotlin.math.abs

@Suppress("unused")
object Day1 : DayBase(1) {
    var leftList = listOf<Int>()
    var rightList = listOf<Int>()

    override fun solve(input: List<String>) {
        // Parsing
        for (line in input) {
            val (left, right) = line.split(Regex("\\s+"))
            leftList += left.toInt()
            rightList += right.toInt()
        }

        // Sorting
        leftList = leftList.sorted()
        rightList = rightList.sorted()

        // Part 1
        var sum = 0
        for ((index, left) in leftList.withIndex()) {
            sum += abs(left - rightList[index])
        }

        setPart1Answer(sum)

        // Part 2
        sum = 0
        for (left in leftList) {
            var occurrences = 0
            for (right in rightList) {
                if (left > right) {
                    continue
                }
                if (left == right) {
                    occurrences++
                    continue
                }
                break
            }
            sum += left * occurrences
        }

        setPart2Answer(sum)
    }
}
